Onion client 14

CamiloGarciaLaRotta opened the pull request:
**Solving issue** #14

- Code in this branch was tested and meets the criteria of the open issue
- Minor changes in `onion_router.py` to avoid using low-level libs and adhere to the no JSON communication standard
- Minor changes in `dummy_dest.py` to reply to setup SYNs and adhere to the no JSON communication standard
- Folder structure was reformatted to homogenize naming conventions

**Logical Next Steps:** 
- Must actually implement directory's `get_path()` to return a valid JSON with path's hostnames and keys
- Must decide on whats the protocol of communication with the directory node. At this moment I send an empty `b'TODO'` when connecting to it, because the directory node doesn't care what the recieved message says.
- Must **DEFINETIVELY** handle teardown and error handling along the path: clossing one node crashes all other nodes in the VC

CamiloGarciaLaRotta committed:
Restructured folder structure

All onion network components are now under a common folder

CamiloGarciaLaRotta committed:
Base implementation of client

Automatically get path, encapsulate setup onions and continously prompt user for input msg.

**BN** Still uses JSOn for standard communication

CamiloGarciaLaRotta committed:
Reformated repo file structure

CamiloGarciaLaRotta committed:
Merged updates from #16

CamiloGarciaLaRotta committed:
Replaced low level socket.send for socket.sendall

Refer to  [StackOverflow](https://stackoverflow.com/questions/34252273/what-is-the-difference-between-socket-send-and-socket-sendall)

CamiloGarciaLaRotta committed:
Stable version of Onion Client

**TODO** clean and polish, implement `directory.py`s get_path, implement `-v` flag on all onion classes to filter output

CamiloGarciaLaRotta committed:
Cleaned logic, added verbose

Onion Client ready to be deployed

All nodes now output status logs only if the `-v` flag is prompted

DestructiveReasoning commented on destination/dummy_dest.py:
```diff
@@ -41,9 +44,13 @@ def main():
+            msg = conn_socket.recv(2048).decode('utf-8')
+            print('Recieved: {}'.format(msg))
+            if msg == 'SYN': 
+                conn_socket.send(b'ACK')
```
Why is this handshake necessary?

DestructiveReasoning commented on destination/dummy_dest.py:
```diff
@@ -41,9 +44,13 @@ def main():
-            conn_socket.send(mod_msg)
+            msg = conn_socket.recv(2048).decode('utf-8')
+            print('Recieved: {}'.format(msg))
+            if msg == 'SYN': 
```
Why is this handshake necessary?

stumash replied:
agree

CamiloGarciaLaRotta replied:
@DestructiveReasoning @stumash 

To clarify on the SYN/ACK:

I had to adhere to the method of `onion_router.py` of establishing the VC.
Because each router returns ACK on the first message and then waits for the second one to establish the next hop connection, I had to choose a dummy string to send as a first message.
Choosing SYN was the natural answer

`onion_router.py` line 63: 
```python
#(TODO:decrypt data, initialize+respond with symkey. for now send 'ACK')
back_conn.sendall("ACK".encode('utf-8'))
```

As to why it was also added to the destination node, it was to keep consistency throughout the configuration of the VC as you can see here:

`onion_client.py` line 102
```python
 def setup_vc(path, conn):
 # get ACK from first node
 conn.sendall(b'SYN')
 response = conn.recv(1024).decode('utf-8').rstrip()
 if response != 'ACK': return false
 if args.verbose: print('ACK recieved')

 # get ACK for remaining nodes and destination
 for i in range(1,4):
     setup_onion = encapsulate(path[i]['addr'], path[i]['key'], 'SYN')
     if args.verbose: print('setup_onion: {}'.format(setup_onion))
     
     # TODO encrypt data
     
     conn.sendall(json.dumps(setup_onion).encode('utf-8'))
     response = conn.recv(1024).decode('utf-8').rstrip()
      if response != 'ACK': return false
     if args.verbose: print('ACK recieved')
```

I am aware it won't allow accessing websites because they won't return `ACK`s, I chose to add because currently, we are only talking to internal nodes. But if issue #15 is expected to be solved before the project submission and thus allowing us to communicate to other ports (e.g. accessing websites) I can change it right now.

I'll wait on your decision

DestructiveReasoning replied:
I don't think this is necessary, because it's just not a realistic enough scenario in my opinion. The originator knows how many intermediate nodes there are, so can't he just wait for that many ACK's?

stumash replied:
I see. You put the handshake here so that you can do the loop in `setup_vc` in `onion_client`.  Also so that all nodes (whether it be directory, router, or destination) respond with ACK first before they then go on to do whatever it is that they do.

However, actual destinations (like, www.google.com, for example) don't follow this protocol - they just respond.  So, maybe it's not such a good idea for the destination to imitate the behaviour of the routers for 'consistency'.  Maybe, `setup_vc` should only initialise communication between the routers and that the first packet from the originator intended for processing by the destination will actually be the first packet that the destination receives from the originator -- as opposed to how in this PR, the destination receives a `SYN` as part of the `setup_vc`.

Do you agree?

CamiloGarciaLaRotta replied:
that works, so to conclude:
- Recieve 3 `ACK`s because 3 intermediate nodes
- For the actual destination first package received is actual first data packet

When you confirm I'll apply the change

stumash replied:
yes cool

DestructiveReasoning replied:
Yeah that's what I meant. Sounds good

CamiloGarciaLaRotta replied:
Fixed in 31e617e7b16ab2a1f0db2407a3bcd6f22e4cab33

DestructiveReasoning commented:
Looks nice. I just don't understand what that SYN/ACK thing is with the `dummy_dest.py`, and likewise the `received SYN` log in `onion_router.py`.

CamiloGarciaLaRotta commented:
Not necessary, the same way the routers returning ACK is not necessary
right now.

I added it to keep consistency throughout the configuration of the VC.

You can see its use in the client's setup_vc()

On Nov 6, 2017 5:16 PM, "Harley Wiltzer" <notifications@github.com> wrote:

> *@DestructiveReasoning* commented on this pull request.
> ------------------------------
>
> In destination/dummy_dest.py
> <https://github.com/CamiloGarciaLaRotta/Cebolla/pull/21#discussion_r149220179>
> :
>
> > @@ -41,9 +44,13 @@ def main():
>  def reply(conn_socket):
>      try:
>          while True:
> -            msg = conn_socket.recv(2048)
> -            mod_msg = msg.upper()
> -            conn_socket.send(mod_msg)
> +            msg = conn_socket.recv(2048).decode('utf-8')
> +            print('Recieved: {}'.format(msg))
> +            if msg == 'SYN':
> +                conn_socket.send(b'ACK')
>
> Why is this handshake necessary?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/CamiloGarciaLaRotta/Cebolla/pull/21#pullrequestreview-74581327>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AQZDutDfSKHnxD8xhL49ZkCJY4b5_FYnks5sz4UfgaJpZM4QTuiB>
> .
>

stumash commented on client/onion_client.py:
```diff
@@ -0,0 +1,145 @@
+# validate args against conditions
+if args.port < 5551 or args.port > 5557: # 7 group members, each get a port
+    parser.error("port must satisfy: 5551 <= port <= 5557")
+if args.verbose: print('Verbosity ON\n')
```
I don't think it's conventional to tell the user via `print` statement that they have in fact passed the `-v` flag.

CamiloGarciaLaRotta replied:
True indeed, fixed in 267e7f2

CamiloGarciaLaRotta commented:
True, I can remove it

On Nov 6, 2017 5:49 PM, "Stuart Mashaal" <notifications@github.com> wrote:

> *@stumash* commented on this pull request.
> ------------------------------
>
> In client/onion_client.py
> <https://github.com/CamiloGarciaLaRotta/Cebolla/pull/21#discussion_r149227573>
> :
>
> > +#   CLI ARGS
> +###############################################################################
> +
> +parser = argparse.ArgumentParser()
> +
> +parser.add_argument("destination", help="hostname to connect to")
> +parser.add_argument("port", help="port to listen on", type=int)
> +parser.add_argument("-v", "--verbose",
> +                    help="level of logging verbose", action="store_true")
> +
> +args = parser.parse_args()
> +
> +# validate args against conditions
> +if args.port < 5551 or args.port > 5557: # 7 group members, each get a port
> +    parser.error("port must satisfy: 5551 <= port <= 5557")
> +if args.verbose: print('Verbosity ON\n')
>
> I don't think it's conventional to tell the user via print statement that
> they have in fact passed the -v flag.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/CamiloGarciaLaRotta/Cebolla/pull/21#pullrequestreview-74589760>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AQZDuh6NsVy81Yw9oyUpvVZCQ2DW_Vkdks5sz4zegaJpZM4QTuiB>
> .
>

CamiloGarciaLaRotta committed:
Removed uneccessary print for -v flag

stumash commented on directory/onion_directory.py:
```diff
@@ -110,14 +109,17 @@ def handle_path_request(conn):
+    return [{'addr':'lab2-1','key':'KEY_1'},
+            {'addr':'lab2-2','key':'KEY_2'},
+            {'addr':'lab2-3','key':'KEY_3'}]
 
```
I don't see why this change was necessary

CamiloGarciaLaRotta replied:
it is not a change, rather a stub based on the specifications for the directory node:
it should return 3 addresses and their keys.
Coding that functionality isn't part of the scope of this PR-issue

stumash replied:
it seems you changed the stub from doing `random.sample(NODES, 3)` to 
```python
return [{'addr':'lab2-1','key':'KEY_1'},
        {'addr':'lab2-2','key':'KEY_2'},
        {'addr':'lab2-3','key':'KEY_3'}]
```

but ya I guess it's just a formatting thing

stumash commented on client/onion_client.py:
```diff
@@ -0,0 +1,144 @@
+        print('\nReply from {}: {}'.format('ADD SRC',msg))
+
+
+def shut_down_directory_node():
```
*`shut_down_client_node` you mean?

CamiloGarciaLaRotta replied:
Fixed in 2dadc05

stumash commented on client/onion_client.py:
```diff
@@ -0,0 +1,144 @@
+import argparse             # for command-line argument parsing
+def shut_down_directory_node():
```
this function is defined twice

CamiloGarciaLaRotta replied:
Fixed in  2dadc05

stumash commented on client/onion_client.py:
```diff
@@ -0,0 +1,144 @@
+
+LISTEN_SOCKET = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+LISTEN_SOCKET.bind((HOST, PORT))
+LISTEN_SOCKET.listen(1)
```
the `LISTEN_SOCKET` is not used, we don't need it

CamiloGarciaLaRotta replied:
Fixed in  2dadc05

stumash commented on directory/onion_directory.py:
```diff
@@ -5,11 +5,10 @@
-
-
+#    - directory ping of nodes causes routers to fail, must start directory
+#      before router
```
this wasn't the case for me

CamiloGarciaLaRotta replied:
it is, when the error occurred I rolled back to your latest change on `onion_router.py` 645c96cca878fa63bf5a07826560de6446f32856

It crashes there too, I assumed that's why you opened issue #19

stumash replied:
embarassing.  We'll have to handle that. nice find.

CamiloGarciaLaRotta committed:
Removed unused variables and functions

stumash reviewed (approved):
lookin good

CamiloGarciaLaRotta committed:
Onion Client no longuer needs ACK from dst

CamiloGarciaLaRotta committed:
Nodes do not clog ports anymore if they crash

By adding the `daemon` flag at thread instantiation time we ensure all ports are closed whenever a program exits

To Complement this addition, I have homogenized all verbose flags under [Status], [Data] and [Error] banners

CamiloGarciaLaRotta committed:
Merged changes from PR #17

Adjusted `onion_client.py` to comply to the new `port` key in the setup onion

Changes were tested succesfully
